presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Memo.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.plan.ChildReplacer;
+import com.facebook.presto.sql.planner.plan.OutputNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+
+public class Memo
+{
+    private final PlanNodeIdAllocator idAllocator;
+    private final int rootGroup;
+
+    private int nextGroupId;
+
+    private final Map<Integer, PlanNode> membership = new HashMap<>();
+    private final Map<Integer, Integer> referenceCounts = new HashMap<>();
```

Comment:
> what is the purpose of this map? Is it used for garbage collection? Then is not mark and sweep a better approach for collection?

Yes, it's for GC. It's used below (see incrementReferenceCounts and decrementReferenceCounts).

Mark & sweep is another alternative (and I considered it), but it raises questions about how often to do it, how big to let the memo grow, etc. Since there can be no cycles in the memo structure, it's also overkill.

> Couldn't you use java weak references to just remove groups for which there is no GroupReference? You would need some sort of GroupId, but that should be fine.

I considered this, too, but I don't think it's that much easier. It poses other problems: GroupId instance identity would matter, so we'd need to make sure they are never created outside of some very controlled scope. More importantly, if someone outside of the Memo held on to one of them, it would cause a memory leak.

> In what case it is good to remove group? Won't this greatly limit your plan search space? E.g: it might be useful to leave group so that some other rules produce even better plan starting from to be-deleted group.

This memo implementation only keeps around one expression per group. So, effectively, it holds on to just one plan. Keeping track of multiple plans will require a different implementation (I've prototyped some of this in my other branch, mentioned in https://github.com/prestodb/presto/wiki/New-Optimizer).